### PR TITLE
Critical fix of gradle build - add protobuf plugin

### PR DESCRIPTION
**Pull request by ar0ne**
### Contains

Fixes of issue #2873
Add protobuf gradle plugin, update version of protoc, update proto files, remove useless tasks, remove redundant files.

### How to test

Change something in *.proto files. 
And then recompile module: `$ ./gradlew engine:recompileProto`
See that related java files were updated.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Can one of the admins please verify this patch?

File: build.gradle [hunk 1]
```diff
@@ -32,6 +32,10 @@ buildscript {
         // Needed for caching reflected data during builds
         classpath 'org.reflections:reflections:0.9.10'
         classpath 'dom4j:dom4j:1.6.1'
+
+        // plugin for compilation Protobuf files
+        // The latest version is 0.8.1, but it requires Gradle 2.12
```
Comment (skaldarnar) [inline build.gradle]: I guess we can (should) update our Gradle wrapper so that we can use latest version.

File: build.gradle [hunk 2]
```diff
@@ -170,18 +176,10 @@ configurations.all {
 clean.doLast {
     new File(dirNatives).deleteDir()
     new File(dirConfigMetrics).deleteDir()
+    new File(dirProtobufGen).deleteDir()
     println "Cleaned root - don't forget to re-extract stuff! 'gradlew extractNatives extractConfig' will do so, or 'gradlew idea' (or eclipse)"
```
Comment (skaldarnar) [inline build.gradle]: You probably want to add a note about re-generating protobuf files here...

File: engine/build.gradle [hunk 3]
```diff
@@ -152,6 +154,9 @@ dependencies {
         println "*** Setting a CrashReporter binary dependency, not present as source"
         compile group: 'org.terasology.crashreporter', name: 'cr-terasology', version: '4.0.1'
     }
+
+    compile 'com.google.protobuf:protobuf-java:3.2.0'
```
Comment (skaldarnar) [inline engine/build.gradle]: This is the same as in l. 96 - remove it.

File: engine/build.gradle [hunk 4]
```diff
@@ -152,6 +154,9 @@ dependencies {
...
+    protobuf files("src/main/protobuf/EntityData.proto", "src/main/protobuf/NetMessage.proto")
```
Comment (skaldarnar) [inline engine/build.gradle]: What is this line for? I thought that the plugin is clever enough to pick up the *.proto files from `src/main/protobuf` automatically. If not, you should specify a new source set for that purpose.

File: engine/build.gradle [hunk 5]
```diff
@@ -178,14 +183,58 @@ task cacheReflections {
...
 }
 
 task cleanReflections(type: Delete) {
     description = 'Cleans the reflection cache. Useful in cases where it has gone stale and needs regeneration.'
     delete sourceSets.main.output.classesDir.toString() + "/reflections.cache"
 }
 
+protobuf {
+
+    generatedFilesBaseDir = "protobuf-gen"
```
Comment (skaldarnar) [inline engine/build.gradle]: I think the usual path to generate files into is `src/gen/java` or similar.. 

See https://github.com/google/protobuf-gradle-plugin#change-where-the-generated-files-are

File: engine/build.gradle [hunk 6]
```diff
@@ -178,14 +183,58 @@ task cacheReflections {
...
+        // task to recompile proto
+        task recompileProto {}
+        recompileProto.dependsOn(deleteGeneratedProto)
+        recompileProto.dependsOn(generateProto)
+
+        // copy generated files to src/ directory
+        generateProto.doLast {
+            copy {
+                from "$rootDir/protobuf-gen"
+                into 'src'
```
Comment (skaldarnar) [inline engine/build.gradle]: Why not generate them into the correct path?

**Comment by skaldarnar** [comment]
@GooeyHub ok to test

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by ar0ne** [comment]
Hi @skaldarnar , thanks for reply and comments. I just wanna say that I tried do not to do a lot of changes in code (it's a pretty new for me project) and save original structure, but if it's okay I'll improve this PR and ask you to review again soon. Thanks!

File: build.gradle (hunk 1 above)
Comment (Cervator) [inline build.gradle]: Updating to 2.12 or even 2.14 probably wouldn't be terrible. Gradle 3 I think might come with some other required changes or we'd need to test for quirks more thoroughly.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by ar0ne** [comment]
Hi @skaldarnar, I redid PR, please take a look when you will have time. Thanks!

**Comment by flo** [comment]
When I checkout this branc and just run gradlew idea I get tons of compile errors due to missing files that were previously checked it but are Iguess now suposed to be generated during building?

@skaldarnar I would think it is better when we don't depend on a gradlew task to run first before we can work with the source in an ID|E. e.g. have the source files checked in as previously.

**Comment by oniatus** [comment]
@ar0ne Can you check if the PR is still required and compatible with the latest version as @flo asked?

**Comment by GooeyHub** [comment]
Uh oh, something went wrong with the build. Need to check on that

**Comment by ar0ne** [comment]
Hey @oniatus and @flo, sorry for the long delay.
I thought we don't need these changes at all. So, is it fine if I'll decline PR?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by ar0ne** [comment]
Hi @Cervator, @skaldarnar, @flo and @oniatus 
I did small changes in this old PR. Honestly, if I remember it correctly we discussed and decided that it's pretty useless idea, and probably gradle was too old and we wanted to updated it then. 
From the other hand, with these last changes we only have task for recompiling proto-classes, and, unfortunately, we still keep auto-generated java classes in git. So, it isn't what I've proposed originally. But, if PR is fine now I OK with that and we could do merge. 
How to test:
1. Change proto-file
2. _./gradlew clean engine:generateProto_
3. Check that java-file has been updated
4. Build project

**Comment by jellysnake** [comment]
@Cervator I've tested this out and it works fine for me. Is it mergable now or is it overshadowed by the gradle changes you've got floating around?

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
@jellysnake this is unrelated to my Gradle stuffs. No problem there.

@ar0ne thanks for keeping up with this one for so long :-)

I've tested it out locally and it appears to work, but to my surprise the protobuf code generation task runs by default, which means any `gradlew build` (every Jenkins engine build) will regenerate the data files and then ship the game that way instead of what's in the repo. Intent was to just have the task available as needed when there's a legit change that needs a regen. Hmm.

There's probably a simple way to decouple it as a task dependency, but that's a little out of my reach for now. Should be able to figure it out sometime.

**Comment by DarkWeird** [comment]
https://stackoverflow.com/questions/47192746/how-can-i-disable-a-task-in-build-gradle

Only append line of code... 
```
protoc.onlyIf {false} 
```
Or running task related

**Pull request closed without merging**